pkg/cache/cache.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/kylelemons/godebug/pretty"
+)
+
+type ResourceCache interface {
+	// Sets the key to the provided value, and generates an update
+	// on the queue the value has changed.
+	Set(key string, value interface{})
+
+	// Gets the value associated with the given key.  Returns nil
+	// if the key is not present.
+	Get(key string) (interface{}, bool)
+
+	// Sets the key to the provided value, but does not generate
+	// and update on the queue ever.
+	// Priming should happen only at the begining and only run 
+	// method should have access to it. Hence a private method.
+	prime(key string, value interface{})
+
+	// Deletes the value identified by the given key from the cache, and
+	// generates an update on the queue if a value was deleted.
+	Delete(key string)
+
+	// Lists the keys currently in the cache.
+	ListKeys() []string
+
+	// Returns true when the cache has synced with the datastore,
+	// false until that point.
+	HasSynced() bool
+
+	// Starts the cache.
+	Run(stopChan chan struct{})
+
+	// Sets kubernetes cache indexer
+	SetIndexer(indexer k8sCache.Indexer)
+}
+
+// CalicoCache implements ResourceCache interface
+type CalicoCache struct {
+	threadSafeCache *cache.Cache
+	workqueue       workqueue.RateLimitingInterface
+	calicoClient    *calicoClient.Client
+	synced          bool 							// Flag set to true when cache is synced with etcd datastore
+	objType			string							// Type of object CalicoCache will hold. E.g. calico profiles, policies etc.
+	k8sCacheIndexer k8sCache.Indexer				// K8S cache indexer used while priming of calicoCache.
+}
+
+func NewCache(queue workqueue.RateLimitingInterface, client *calicoClient.Client, objectType string) ResourceCache {
+	
+	return &CalicoCache{
+		threadSafeCache: cache.New(cache.NoExpiration, cache.DefaultExpiration),
+		workqueue:       queue,
+		calicoClient:    client,
+		synced:			 false,
+		objType:		 objectType,
+	}
+}
+
+func (c *CalicoCache) Set(key string, obj interface{}) {
+
+	
+	// Converting obj to calico representation
+	profile := api.Profile{
+		TypeMetadata: unversioned.TypeMetadata{
+			Kind:  "profile",     
+			APIVersion: "v1",
+		}, 
+		Metadata: api.ProfileMetadata{
+			Name: obj.(*v1.Namespace).Name,
+		},
+	}
+	
+	if existingProfile, found := c.threadSafeCache.Get(key); found {
+
+		glog.Infof("%s found in cache. comparing", existingProfile)
+		diff := pretty.Compare(profile,existingProfile)
+		glog.Info("Diff:",diff)
+
+		//reflect.DeepEqual() is not able to correctly compare the profile objects
+		if len(diff)!=0 {
+			glog.Infof("%s and %s do not match.", profile, existingProfile)
+			
+			c.threadSafeCache.Set(key, profile, cache.NoExpiration)
+			c.workqueue.Add(key)
+
+		} else {
+			// Nothing to do. Object already exists in cache
+		}
+	} else {
+		glog.Infof("%s not found in cache. Adding in cache", obj)
+	
+		c.threadSafeCache.Set(key, profile, cache.NoExpiration)
+		c.workqueue.Add(key)
+	}
+}
+
+func (c *CalicoCache) Delete(key string) {
+
```

Comment:
formatting nit: we should remove all of these newlines at the start of functions.